fast_dqn/deep_deterministic_policy_gradient.py

```diff
@@ -0,0 +1,139 @@
+        self._main_critic = make_critic()
+        self._actor_vars = self._main_actor.trainable_variables
+        self._critic_vars = self._main_critic.trainable_variables
+        self._target_actor = make_actor()
+        self._target_critic = make_critic()
+        self._exec_actor = make_actor()
+        self._exec_critic = make_critic()
+
+    def _preprocess_states(self, states):
+        return tf.cast(states, tf.float32) / 255.0
+
+    # TODO: naming is a little confusing... this function only gets
+    # Called to update the critic network
+    @tf.function
+    def predict_actions_and_values(self, states, network):
+        # Critic does not update encoder
+        encoding = tf.stop_gradient(self._encoder(states))
+        actions = {
+            'main': self._main_actor,
+            'target': self._target_actor,
+            'exec': self._exec_actor,
+        }[network](encoding)
+        actions = tf.clip_by_value(actions, -self.action_limit, self.action_limit)
+        values = {
+            'main': self._main_critic,
+            'target': self._target_critic,
+            'exec': self._exec_actor,
+        }[network](tf.concat([encoding, actions], axis=1))
+        return actions, values
+
+    @tf.function
+    def predict_actions(self, states, network):
+        # Critic does not update encoder
+        encoding = tf.stop_gradient(self._encoder(states))
+        return {
+            'main': self._main_actor,
+            'target': self._target_actor,
+            'exec': self._exec_actor,
+        }[network](encoding)
+
+    @tf.function
+    def predict_values(self, states, actions, network):
+        encoding = self._encoder(states)
+        concat = tf.concat([encoding, actions], axis=1)
+        return {
+            'main': self._main_critic,
+            'target': self._target_critic,
+            'exec': self._exec_critic,
+        }[network](concat)
+
+    @tf.function
+    def train(self, states, actions, rewards, next_states, dones):
+        states, next_states = self._preprocess_states(states), self._preprocess_states(next_states)
+        next_actions = tf.clip_by_value(self.predict_actions(next_states, network='target'), -self.action_limit, self.action_limit)
+        next_Q = self.predict_values(next_states, next_actions, network='target')
+
+        done_mask = 1.0 - tf.cast(dones, tf.float32)
+        targets = rewards + self.discount * done_mask * next_Q
+
+        # Critic update
+        with tf.GradientTape() as tape:
+            Q = self.predict_values(states, actions, network='main')
+            # TODO: Should we use huber loss here?
+            loss = tf.reduce_mean(huber_loss(targets - Q))
+
+        gradients = tape.gradient(loss, self._critic_vars)
+        self.critic_optimizer.apply_gradients(zip(gradients, self._critic_vars))
+
+        # Actor update
+        with tf.GradientTape() as tape:
+            _, Q = self.predict_actions_and_values(states, network='main')
```

Comment:
Let's share the forward pass for actor + critic when training, for computational efficiency.